### ISSUE TITLE
`lein deps :verify` reports everything as :unsigned

### ISSUE THREAD

**Issue by miikka** [technomancy/leiningen#2697]
I've tried to run `lein deps :verify` against a couple of projects and it reports all dependencies as `:unsigned`. I think it's failing to download the signatures again like in #2567.

**To Reproduce**
I pushed a new signed jar on Clojars to double-check that I'm verifying a signed dep. It's at `[org.clojars.miikka/signed-artifact "0.1.0"]`. You can see at https://repo.clojars.org/org/clojars/miikka/signed-artifact/0.1.0/ that there are signature `.asc` files there.

```sh
git clone https://github.com/miikka/weird-artifacts
cd weird-artifacts/has-dependencies
lein deps :verify
```

**Actual behavior**
On the system where I made and signed the artifact using `lein release`, it works as expected. This is probably because the `.asc` files exist in `~/.m2/repository/org/clojars/miikka/signed-artifact/0.1.0`.  However, on a system with empty `~/.m2/repository` cache, this happens:

```
# lein deps :verify
Retrieving org/clojars/miikka/signed-artifact/0.1.0/signed-artifact-0.1.0.pom from clojars
Retrieving org/clojars/miikka/signed-artifact/0.1.0/signed-artifact-0.1.0.jar from clojars
:unsigned [clojure-complete "0.2.5" :exclusions [[org.clojure/clojure]]]
:unsigned [nrepl "0.6.0" :exclusions [[org.clojure/clojure]]]
:unsigned [org.clojars.miikka/signed-artifact "0.1.0"]
:unsigned [org.clojure/clojure "1.10.1"]
:unsigned [org.clojure/core.specs.alpha "0.2.44"]
:unsigned [org.clojure/spec.alpha "0.2.176"]
```

**Expected behavior**
I expected that `org.clojars.miikka/signed-artifact` would be `:signed` or `:no-key` (or there would be some error message).

**Link to sample project**
https://github.com/miikka/weird-artifacts/blob/master/has-dependencies/project.clj

**Environment**
- Leiningen Version: 2.9.4
- JDK Version: openjdk version "11.0.8" 2020-07-14
- OS: `clojure:latest` Docker image (Debian)

**Comment by glts** [comment]
Here is the output of `lein deps :verify` for your project on my machine:

```
gpg: assuming signed data in '/home/glts/.m2/repository/clojure-complete/clojure-complete/0.2.5/clojure-complete-0.2.5.jar'
gpg: Signature made So 04 Feb 2018 22:18:44 CET
gpg:                using RSA key 67E57A31E9E02368EEAC5CCFBB1DBD3616F1DDB9
gpg:                issuer "trptcolin@gmail.com"
gpg: Good signature from "Colin Jones <trptcolin@gmail.com>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: 67E5 7A31 E9E0 2368 EEAC  5CCF BB1D BD36 16F1 DDB9
:signed [clojure-complete "0.2.5" :exclusions [[org.clojure/clojure]]]
:unsigned [nrepl "0.7.0" :exclusions [[org.clojure/clojure]]]
gpg: assuming signed data in '/home/glts/.m2/repository/org/clojars/miikka/signed-artifact/0.1.0/signed-artifact-0.1.0.jar'
gpg: Signature made Mo 27 Jul 2020 19:41:09 CEST
gpg:                using RSA key 2A6BD325C578BFB4FA31FC6CF596099EEF199C32
gpg: Can't check signature: No public key
gpg: keyserver receive failed: No data
:no-key [org.clojars.miikka/signed-artifact "0.1.0"]
:unsigned [org.clojure/clojure "1.10.1"]
:unsigned [org.clojure/core.specs.alpha "0.2.44"]
:unsigned [org.clojure/spec.alpha "0.2.176"]
```

After adding `:checksum :ignore` in project.clj this becomes (now omitting the lines prefixed `gpg:`):

```
lein deps :verify
:signed [clojure-complete "0.2.5" :exclusions [[org.clojure/clojure]]]
:unsigned [nrepl "0.7.0" :exclusions [[org.clojure/clojure]]]
:no-key [org.clojars.miikka/signed-artifact "0.1.0"]
:signed [org.clojure/clojure "1.10.1"]
:signed [org.clojure/core.specs.alpha "0.2.44"]
:signed [org.clojure/spec.alpha "0.2.176"]
```

**This result is correct.** (Incidentally, as you can see, you haven’t published your public key to a key server accessible in my setup. I use key server hkps://keys.openpgp.org, the current Ubuntu default.)

The full explanation is complicated. Very briefly, `lein deps :verify` does work, but there are major subtleties that sabotage the default experience. Among other things, these are to do with certain Maven defaults and GnuPG defaults. Someone should take the time to revise the documentation …

**Comment by glts** [comment]
Quick top-of-my-mind listing of some of the tricky bits:

* GnuPG recently disabled the `auto-key-retrieve` setting by default. For verification you obviously need to retrieve the public keys, this is done by adding a line `auto-key-retrieve` in ~/.gnupg/gpg.conf (followed by `gpgconf --reload`).
* PGP recently suffered an attack on public key servers, so some distros (Debian, Ubuntu) changed the default key servers to an implementation where keys cannot be uploaded simply with `gpg --send-keys`. You have to visit https://keys.openpgp.org to publish your keys.
* Maven (not Leiningen/Pomegranate!) has an odd default where checksum files are not deployed by default. Thus you will find that dependencies from Maven central have a failing checksum check and appear as unsigned. Thus for Maven dependencies (org.clojure in your example project) you need to specify `:checksum :ignore` to ignore the failing checksums.

I spent a bit of time on this and improved a few things a year ago. I want to have this work well – it is valuable. But not currently working on it. Cheers.

**Comment by miikka** [comment]
@gits Your output shows my deps as `:no-key` and that's indeed correct. But why does my output show it as `:unsigned`?

**Comment by glts** [comment]
@miikka works for me. using Ubuntu 20.04, Leiningen 2.9.4 on Java 11.0.8 OpenJDK 64-Bit Server VM, GnuPG 2.2.19

**Comment by jessesaga** [comment]
Hi - I stumbled across this issue and am experiencing the same behavior. I have my own package (using GNU Guix) built from the 2.9.4 release and :verify always fails. I'm working around the problem by modifying the lein-pkg script during packaging by adding....

-Daether.checksums.forSignature=true

...to the list of JVM parameters in the script.

**Comment by glts** [comment]
@jessesaga this setting should already be the default in Leiningen/Pomegranate (as far as I remember).

Hoping someone can provide complete reproduction steps – the above works for me, with the modifications I listed as ‘tricky bits’.

**Comment by miikka** [comment]
A bit cumbersome but portable way to reproduce this is using the [clojure docker image](https://hub.docker.com/_/clojure):

```sh
docker pull clojure:latest
docker run -it clojure:latest /bin/bash
# The following commands are run inside Docker
apt-get update && apt-get install -y git gpg
git clone https://github.com/miikka/weird-artifacts
cd weird-artifacts/has-dependencies
lein deps :verify
```

The versions:

```
Leiningen 2.9.3 on Java 11.0.8 OpenJDK 64-Bit Server VM
gpg (GnuPG) 2.2.12
```

**Comment by glts** [comment]
@miikka And you did try adding `:checksum :ignore`, right

**Comment by jessesaga** [comment]
@glts I can confirm that adding :checksum :ignore to project.clj changes the behavior and lein deps :verify starts reporting some signing related messages as opposed to just silently failing and reporting everything as unsigned. This was validated using the 2.9.4 release downloaded from Github in conjunction with the 2.9.4 lein-pkg script. (incidentally the 2.9.4 lein-pkg script has an unterminated string that has to be fixed before it will run but it appears this has been fixed in master).

At this point, I don't know which workaround I should run with. Perhaps I don't understand the behavior of the :checksum :ignore directive, but at face value you'd think it would be less secure and not validate signatures. If it's causing the exact opposite to happen ....... it's not very intuitive.

If I keep my workaround patch that adds -Daether.checksums.forSignature=true to the lein script, I don't need to alter every project.clj. Doesn't that seem like a better approach or am I missing something?

BTW - in order to replicate the failure on projects without the :checksum :ignore directive, the following steps work for me

- start with clean ~/.gnupg and ~/.m2 directories
- download the leiningen-2.9.4-standalone.zip release from github
- download the lein-pkg script from the 2.9.4 tagged github tree
- rename  leiningen-2.9.4-standalone.zip to leiningen-2.9.4-standalone.jar
- edit lein-pkg and fix the unterminated string issue (around line # 88)
- edit lein-pkg to reference leiningen-2.9.4-standalone.jar 
- run lein-pkg deps :verify in any normal clojure project that doesn't have the :checksum :ignore directive

```
$ ./lein-pkg deps :verify
Retrieving org/clojure/clojure/1.10.1/clojure-1.10.1.pom from central
Retrieving org/clojure/spec.alpha/0.2.176/spec.alpha-0.2.176.pom from central
Retrieving org/clojure/pom.contrib/0.2.2/pom.contrib-0.2.2.pom from central
Retrieving org/clojure/core.specs.alpha/0.2.44/core.specs.alpha-0.2.44.pom from central
Retrieving nrepl/nrepl/0.6.0/nrepl-0.6.0.pom from clojars
Retrieving clojure-complete/clojure-complete/0.2.5/clojure-complete-0.2.5.pom from clojars
Retrieving org/clojure/spec.alpha/0.2.176/spec.alpha-0.2.176.jar from central
Retrieving org/clojure/core.specs.alpha/0.2.44/core.specs.alpha-0.2.44.jar from central
Retrieving org/clojure/clojure/1.10.1/clojure-1.10.1.jar from central
Retrieving nrepl/nrepl/0.6.0/nrepl-0.6.0.jar from clojars
Retrieving clojure-complete/clojure-complete/0.2.5/clojure-complete-0.2.5.jar from clojars
:unsigned [clojure-complete "0.2.5" :exclusions [[org.clojure/clojure]]]
:unsigned [nrepl "0.6.0" :exclusions [[org.clojure/clojure]]]
:unsigned [org.clojure/clojure "1.10.1"]
:unsigned [org.clojure/core.specs.alpha "0.2.44"]
:unsigned [org.clojure/spec.alpha "0.2.176"]
```

Removing .m2 directory and adding :checksum :ignore to the project.clj results in the following output.
```
 $ ./lein-pkg deps :verify
Retrieving org/clojure/clojure/1.10.1/clojure-1.10.1.pom from central
Retrieving org/clojure/spec.alpha/0.2.176/spec.alpha-0.2.176.pom from central
Retrieving org/clojure/pom.contrib/0.2.2/pom.contrib-0.2.2.pom from central
Retrieving org/clojure/core.specs.alpha/0.2.44/core.specs.alpha-0.2.44.pom from central
Retrieving nrepl/nrepl/0.6.0/nrepl-0.6.0.pom from clojars
Retrieving clojure-complete/clojure-complete/0.2.5/clojure-complete-0.2.5.pom from clojars
Retrieving org/clojure/core.specs.alpha/0.2.44/core.specs.alpha-0.2.44.jar from central
Retrieving org/clojure/spec.alpha/0.2.176/spec.alpha-0.2.176.jar from central
Retrieving org/clojure/clojure/1.10.1/clojure-1.10.1.jar from central
Retrieving nrepl/nrepl/0.6.0/nrepl-0.6.0.jar from clojars
Retrieving clojure-complete/clojure-complete/0.2.5/clojure-complete-0.2.5.jar from clojars
gpg: assuming signed data in '/home/jesse/.m2/repository/clojure-complete/clojure-complete/0.2.5/clojure-complete-0.2.5.jar'
gpg: Signature made Sun 04 Feb 2018 04:18:44 PM EST
gpg:                using RSA key 67E57A31E9E02368EEAC5CCFBB1DBD3616F1DDB9
gpg:                issuer "trptcolin@gmail.com"
gpg: Can't check signature: No public key
gpg: key BB1DBD3616F1DDB9: public key "Colin Jones <trptcolin@gmail.com>" imported
gpg: Total number processed: 1
gpg:               imported: 1
gpg: assuming signed data in '/home/jesse/.m2/repository/clojure-complete/clojure-complete/0.2.5/clojure-complete-0.2.5.jar'
gpg: Signature made Sun 04 Feb 2018 04:18:44 PM EST
gpg:                using RSA key 67E57A31E9E02368EEAC5CCFBB1DBD3616F1DDB9
gpg:                issuer "trptcolin@gmail.com"
gpg: Good signature from "Colin Jones <trptcolin@gmail.com>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: 67E5 7A31 E9E0 2368 EEAC  5CCF BB1D BD36 16F1 DDB9
:signed [clojure-complete "0.2.5" :exclusions [[org.clojure/clojure]]]
:unsigned [nrepl "0.6.0" :exclusions [[org.clojure/clojure]]]
:unsigned [org.clojure/clojure "1.10.1"]
:unsigned [org.clojure/core.specs.alpha "0.2.44"]
:unsigned [org.clojure/spec.alpha "0.2.176"]
```

Removing :checksum :ignore and adding -Daether.checksums.forSignature=true to the lein-pkg script results in the following identical output.
```
$ ./lein-pkg deps :verify
Retrieving org/clojure/clojure/1.10.1/clojure-1.10.1.pom from central
Retrieving org/clojure/spec.alpha/0.2.176/spec.alpha-0.2.176.pom from central
Retrieving org/clojure/pom.contrib/0.2.2/pom.contrib-0.2.2.pom from central
Retrieving org/clojure/core.specs.alpha/0.2.44/core.specs.alpha-0.2.44.pom from central
Retrieving nrepl/nrepl/0.6.0/nrepl-0.6.0.pom from clojars
Retrieving clojure-complete/clojure-complete/0.2.5/clojure-complete-0.2.5.pom from clojars
Retrieving org/clojure/spec.alpha/0.2.176/spec.alpha-0.2.176.jar from central
Retrieving org/clojure/core.specs.alpha/0.2.44/core.specs.alpha-0.2.44.jar from central
Retrieving org/clojure/clojure/1.10.1/clojure-1.10.1.jar from central
Retrieving nrepl/nrepl/0.6.0/nrepl-0.6.0.jar from clojars
Retrieving clojure-complete/clojure-complete/0.2.5/clojure-complete-0.2.5.jar from clojars
gpg: assuming signed data in '/home/jesse/.m2/repository/clojure-complete/clojure-complete/0.2.5/clojure-complete-0.2.5.jar'
gpg: Signature made Sun 04 Feb 2018 04:18:44 PM EST
gpg:                using RSA key 67E57A31E9E02368EEAC5CCFBB1DBD3616F1DDB9
gpg:                issuer "trptcolin@gmail.com"
gpg: Can't check signature: No public key
gpg: key BB1DBD3616F1DDB9: public key "Colin Jones <trptcolin@gmail.com>" imported
gpg: Total number processed: 1
gpg:               imported: 1
gpg: assuming signed data in '/home/jesse/.m2/repository/clojure-complete/clojure-complete/0.2.5/clojure-complete-0.2.5.jar'
gpg: Signature made Sun 04 Feb 2018 04:18:44 PM EST
gpg:                using RSA key 67E57A31E9E02368EEAC5CCFBB1DBD3616F1DDB9
gpg:                issuer "trptcolin@gmail.com"
gpg: Good signature from "Colin Jones <trptcolin@gmail.com>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: 67E5 7A31 E9E0 2368 EEAC  5CCF BB1D BD36 16F1 DDB9
:signed [clojure-complete "0.2.5" :exclusions [[org.clojure/clojure]]]
:unsigned [nrepl "0.6.0" :exclusions [[org.clojure/clojure]]]
:unsigned [org.clojure/clojure "1.10.1"]
:unsigned [org.clojure/core.specs.alpha "0.2.44"]
:unsigned [org.clojure/spec.alpha "0.2.176"]
```

**Comment by glts** [comment]
Can you also repro with a regular `lein` installation? Frankly I don't have the energy to debug this odd lein-pkg repro right now, sorry.

As for the motivation for the `:checksum :ignore` directive, I have tried to explain this briefly above. There is a difference in checksum handling (resolution/deployment) in Maven versus Leiningen. Have a look at any dependency in a Maven repository versus on Clojars:

* https://repo1.maven.org/maven2/org/clojure/spec.alpha/0.2.176/
* https://repo.clojars.org/ch/gluet/couplet/0.2.1/

With Leiningen/Clojars we resolve/deploy checksums for *all* files. Maven on the other hand omits checksum files for signatures, which is why `:checksum :ignore` is necessary to work around this suboptimal default, but only for dependencies in Maven repos. For dependencies on Clojars this setting is not necessary. Hope this helps.

**Comment by jessesaga** [comment]
Thank you. I think I understand the motivation for the :checksum keyword now and am able to achieve desirable results (all dependencies get validated against their signatures by default by overriding the default repositories in all of my project.clj files.

```
  :repositories ^:replace [["central" {:url "https://repo1.maven.org/maven2/" :snapshots false :checksum :ignore}]
                           ["clojars" {:url "https://repo.clojars.org/"}]]
```

Would it make sense to set this as a more sensible default for default-projects in leiningen? 
```
(def default-repositories
  (with-meta
    [["central" {:url "https://repo1.maven.org/maven2/" :snapshots false :checksum :ignore }]
     ["clojars" {:url "https://repo.clojars.org/"}]]
    {:reduce reduce-repo-step}))
```

In terms of the original issue - 'everything is reported as :unsigned' - I'm out of time to troubleshoot as well. Between the repository workaround and my script patch, I've achieved my goal to vet the dependency chains for all of my projects.

**Comment by glts** [comment]
As is apparent from https://github.com/technomancy/leiningen/issues/2697#issuecomment-673955034, this bug report concerns an older version of Leiningen (2.9.3), the fix was released in 2.9.4.

**Comment by miikka** [comment]
@gits Hey, that seems to be the case! Sorry for mixing up the versions. What was the fix? Looking at the diff between 2.9.3 and 2.9.4 it's not immediately obvious to me.

**Comment by glts** [comment]
Yes, it is not obvious because the fix for #2567 landed in master in time for Leiningen 2.9.2, but there was a problem during the release process for both Leiningen 2.9.2 and 2.9.3, so the fix was finally made available only in 2.9.4.
